File: algopt/rebalancer/examples/website/solvers/strategy_examples.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Solver Strategy Examples
 *
 * This file demonstrates solver strategy patterns shown in the
 * solver documentation.
 */

// example_start
#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

// Helper to set up a minimal problem for solver demonstrations
void setupMinimalProblem(ProblemSolver& solver) {
  solver.setObjectName("task");
  solver.setContainerName("host");

  std::map<std::string, std::vector<std::string>> assignment;
  assignment["host0"] = {"task0", "task1", "task2", "task3"};
  assignment["host1"] = {};
  assignment["host2"] = {};
  solver.setAssignment(assignment);

  std::map<std::string, double> cpu;
  cpu["task0"] = 1.0;
  cpu["task1"] = 2.0;
  cpu["task2"] = 1.5;
  cpu["task3"] = 2.5;
  solver.addObjectDimension("cpu", std::move(cpu));

  std::map<std::string, double> memory;
  memory["task0"] = 4.0;
  memory["task1"] = 3.0;
  memory["task2"] = 5.0;
  memory["task3"] = 2.0;
  solver.addObjectDimension("memory", std::move(memory));

  std::map<std::string, double> data_size;
  data_size["task0"] = 10.0;
  data_size["task1"] = 20.0;
  data_size["task2"] = 15.0;
  data_size["task3"] = 25.0;
  solver.addObjectDimension("data_size", std::move(data_size));
}

void sequential_solvers() {
  /**
   * Use multiple solvers sequentially.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  setupMinimalProblem(solver);

  // sequential_solvers_start
  // Strategy: Local Search warmup, then Optimal refinement
  LocalSearchSolverSpec lsSpec;
  lsSpec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  lsSpec.solveTime() = 10;
  solver.addSolver(lsSpec);

  OptimalSolverSpec optSpec;
  optSpec.solverPackage() = OptimalSolverPackage::HIGHS;
  optSpec.solveTime() = 60;
  solver.addSolver(optSpec);

  // Local Search runs first, finds good solution
  // Optimal uses it as warmstart and refines
  // sequential_solvers_end
}

void time_limited_approach() {
  /**
   * Time-limited solving approach.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  setupMinimalProblem(solver);

  // time_limited_approach_start
  // Give Local Search 30 seconds to find good solution
  LocalSearchSolverSpec lsSpec;
  lsSpec.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  lsSpec.solveTime() = 30;
  solver.addSolver(lsSpec);

  // Then give Optimal up to 5 minutes to refine
  OptimalSolverSpec optSpec;
  optSpec.solverPackage() = OptimalSolverPackage::HIGHS;
  optSpec.solveTime() = 300;
  solver.addSolver(optSpec);
  // time_limited_approach_end
}

void goal_prioritization() {
  /**
   * Prioritize goals with different weights.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  setupMinimalProblem(solver);

  // goal_prioritization_start
  // High priority: balance load
  BalanceSpec balanceSpec;
  balanceSpec.name() = "balance";
  balanceSpec.scope() = "host";
  balanceSpec.dimension() = "cpu";
  solver.addGoal(std::move(balanceSpec), 10.0);

  // Low priority: minimize movement
  MinimizeMovementSpec moveSpec;
  moveSpec.name() = "minimize-moves";
  moveSpec.scope() = "host";
  moveSpec.dimension() = "data_size";
  solver.addGoal(std::move(moveSpec), 1.0);
  // goal_prioritization_end
}

void constraint_vs_goal() {
  /**
   * Use constraints for hard requirements, goals for preferences.
   */
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  setupMinimalProblem(solver);

  // constraint_vs_goal_start
  // Hard constraint: don't exceed capacity
  CapacitySpec capacitySpec;
  capacitySpec.name() = "capacity";
  capacitySpec.scope() = "host";
  capacitySpec.dimension() = "memory";
  solver.addConstraint(std::move(capacitySpec));

  // Soft goal: prefer balanced CPU
  BalanceSpec balanceSpec;
  balanceSpec.name() = "balance-cpu";
  balanceSpec.scope() = "host";
  balanceSpec.dimension() = "cpu";
  solver.addGoal(std::move(balanceSpec), 1.0);
  // constraint_vs_goal_end
}

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  std::cout << "Running all Solver Strategy examples...\n\n";

  std::cout << "1. Sequential Solvers...\n";
  sequential_solvers();

  std::cout << "\n✓ All Solver Strategy examples completed successfully!\n";
  return 0;
}
// example_end

File: algopt/rebalancer/examples/website/solvers/strategy_examples.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Solver Strategy Examples

This file demonstrates solver strategy patterns shown in the
solver documentation.
"""

# example_start
# pyre-strict
from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    ConstraintSpecs,
    GoalSpecs,
    SolverSpecs,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import (
    BalanceSpec,
    CapacitySpec,
    MinimizeMovementSpec,
)
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    LocalSearchSolverSpec,
    OptimalSolverPackage,
    OptimalSolverSpec,
)


def sequential_solvers():
    """Use multiple solvers sequentially."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # sequential_solvers_start
    # Strategy: Local Search warmup, then Optimal refinement
    solver.addSolver(
        SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(timeLimitMs=10000))
    )
    solver.addSolver(
        SolverSpecs(
            optimalSolverSpec=OptimalSolverSpec(
                solverPackage=OptimalSolverPackage.HIGHS,
                timeLimitMs=60000,
            )
        )
    )

    # Local Search runs first, finds good solution
    # Optimal uses it as warmstart and refines
    # sequential_solvers_end


def time_limited_approach():
    """Time-limited solving approach."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # time_limited_approach_start
    # Give Local Search 30 seconds to find good solution
    solver.addSolver(
        SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(timeLimitMs=30000))
    )

    # Then give Optimal up to 5 minutes to refine
    solver.addSolver(
        SolverSpecs(
            optimalSolverSpec=OptimalSolverSpec(
                solverPackage=OptimalSolverPackage.HIGHS,
                timeLimitMs=300000,
                mipGap=0.05,  # Accept 5% gap
            )
        )
    )
    # time_limited_approach_end


def goal_prioritization():
    """Prioritize goals with different weights."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # goal_prioritization_start
    # High priority: balance load
    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(name="balance", scope="host", dimension="cpu")
        ),
        weight=10.0,
    )

    # Low priority: minimize movement
    solver.addGoal(
        GoalSpecs(
            minimizeMovementSpec=MinimizeMovementSpec(
                name="minimize-moves", scope="host", dimension="data_size"
            ),
        ),
        weight=1.0,
    )
    # goal_prioritization_end


def constraint_vs_goal():
    """Use constraints for hard requirements, goals for preferences."""
    solver = ProblemSolver(service_name="example", service_scope="test")

    # constraint_vs_goal_start
    # Hard constraint: don't exceed capacity
    solver.addConstraint(
        ConstraintSpecs(
            capacitySpec=CapacitySpec(name="capacity", scope="host", dimension="memory")
        )
    )

    # Soft goal: prefer balanced CPU
    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(name="balance-cpu", scope="host", dimension="cpu")
        ),
        weight=1.0,
    )
    # constraint_vs_goal_end


if __name__ == "__main__":
    print("Running all Solver Strategy examples...\n")

    print("1. Sequential Solvers...")
    sequential_solvers()

    print("\n✓ All Solver Strategy examples completed successfully!")
# example_end
